[READY]Modular computer based crew records

atlantiscze opened the pull request:
- Adds a full replacement for old crew records.
- Instead of having multiple split records, a new "crew member record" exists. This record has five categories (generic, medical, security, employment and antag)
- Everyone can view generic fields. Relevant departmental access is required to view med/sec/empl records. Emagged device is required to view antagonist records.
- Record fields are accessed using macro-generated getter/setter methods. I hope this will reduce the chance of field being removed breaking things in the future (as removing a field without removing it's reference would result in compile error, which is easier to catch).
- Generic Med/Sec/Empl/Antag record fields now support paper tag HTML formatting.

Be warned, this is still highly WIP and many things are yet to be done. Opening this so people may take a look at the already (more or less) finished parts or provide suggestions in advance.

TODO list:
- [x] Adding new records
- [x] Better formatting for records list (i'd like to make it crew manifest style)
- [x] Implement record searching
- [x] Implement record printing
- [x] Remove datacore and old records. Redirect everything to use the new records system
- [x] Remove old record viewing computers and replace them with modular devices.
- [x] Test thoroughly
- [x] Changelog entry

atlantiscze committed:
Modular computer based crew records (initial commit)
- Adds a full replacement for old crew records.
- Instead of having multiple split records, a new "crew member record" exists. This record has five categories (generic, medical, security, employment and antag)
- Everyone can view generic fields. Relevant departmental access is required to view med/sec/empl records. Emagged device is required to view antagonist records.
- Record fields are accessed using macro-generated getter/setter methods. I hope this will reduce the chance of field being removed breaking things in the future (as removing a field without removing it's reference would result in compile error, which is easier to catch).

Minijar commented:
On the current medical records, there's sections for allergies and prescriptions but there's no fields in the player settings. Perhaps, have predefined sections to fill out, so everyone's record is a bit more uniform?

Crushtoe commented:
Antag records are still visible via uplink, I assume?

PsiOmegaDelta commented:
I wouldn't add allergies and so forth until they can have actual in-game effects.

atlantiscze commented:
Right now all these unused fields that were never editable by the player, like allergies, are not present in the new records. They can be pretty easily added in the future, if they have an actual mechanical background to them.

For character fluff reasons, you can put this info in the custom record (as people do it nowadays anyway), with an added bonus of support of paper tag HTML formatting, which should make it possible to have a bit nicer records than before.

As for antag records, i am not sure whether it will be comfortably possible to have these available via the uplink, given PDAs are not modular devices and are pretty disgusting code-wise. I'll try to figure out a way that will keep easy access to the records for antags.

mustafakalash commented:
Still working on this?

atlantiscze commented:
Yes, just ran into some complications regarding crew photos. I'll see if i can speed it up a bit.

atlantiscze committed:
Eradicates Datacore (DANGER)
- Gets rid of datacore, reroutes most things to use new records or legacy record made by define
- Various improvements to the records in general, adds searching, New button, etc.
- Gets rid of old types, replaces with placeholder spawners for relevant computers. These will be replaced once actual mapping changes are made, and are in place to ensure the code compiles correctly.

atlantiscze commented:
Rejoice, most of the hard stuff has been finished. Datacore is no longer among us, may it rest in hell.

atlantiscze committed:
Additional work
- Changelog entry
- Printing - moved the relevant method from supply program to nano module and used it.

atlantiscze commented:
It's finally ready for review. I've done some local testing and main parts seem to be working as expected.
![It's happening](http://bit.ly/2u6OjM9)

PurpleMartinJCK commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -277,11 +275,9 @@
-						medical = R.fields["p_stat"]
+		for (var/datum/computer_file/crew_record/E in GLOB.all_crew_records)
+			if (E.GetName() == perpname)
+				medical = E.GetStatus()
```
So does this only give a status now rather than all the info?

atlantiscze replied:
This is the field equivalent to old R.fields["p_stat"] - it's the "Active" "SSD" etc. field.

PurpleMartinJCK replied:
Is that the only medical HUD thing that shows now?

atlantiscze replied:
The "View" button should show more details (name, sex, blood type and a printout of the medical record field)

PurpleMartinJCK commented on code/modules/modular_computers/file_system/programs/generic/records.dm:
```diff
@@ -0,0 +1,327 @@
+	var/datum/computer_file/crew_record/active_record
+	var/message = null
+
+/datum/nano_module/records/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = 1, state = GLOB.default_state)
```
So with the old system, if you wanted to add a new piece to the records, you had to change things in about 4 or 5 places. How many places do you need to add it with this system?

atlantiscze replied:
A new piece, as in, a new field?
You'd have to add one GETTER_SETTER macro line to crew_record.dm which technically declares the field.

If you want the field to be editable/visible in the program you need to add it here, as well as update the template to show it accordingly (as with any other nanoui UI).

If you want it to be visible through any of the HUD examines, you'd have to add it there.

If you want to use it anywhere else, you logically have to add it there as well.

atlantiscze commented:
Due to pretty large amount of files affected, it'd be great if this could be reviewed in advance and then merged quickly once i resolve the conflicts, as they will inevitably pop up again very quickly due to previously mentioned reason.

atlantiscze commented:
@FTangSteve So, are you requesting any changes, or can i dismiss your review now?

Also my request for review still applies, given the amount of affected files, this PR is going to have conflicts again very soon after i resolve them - it would be highly appreciated if it could be reviewed in advance and then merged quickly once i resolve the conflicts.

BlueNexus commented:
At Atlantis's request, I'm going to review this in advance.

BlueNexus reviewed (dismissed):
Looks fine, apart from all the out-of-scope copypaste, carried over from the original system. I'd appreciate someone else reviewing this as well, though.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2017-08-14-ShinyNewRecords

atlantiscze commented:
Okay, conflicts resolved and ready to merge
@BlueNexus Your review was dismissed in the process, might have to make it again, though feature-wise nothing should have changed aside for a small formatting fix for a little thing i've noticed.

comma commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -109,13 +109,6 @@ var/global/datum/matchmaker/matchmaker = new()
-		var/datum/data/record/R2 = find_general_record("name", other.holder.name)
-		var/info = prob(60) ? get_desc_string() : "[holder] and [other.holder] know each other, but the exact nature of their relationship is unclear."
-		if(R1)
-			R1.fields["connections"] |= info
```
Is this readded anywhere?

atlantiscze replied:
No, these are not in the records anymore, but matchmaking should work even w/o this.

comma replied:
Any particular reason they were removed from records?

atlantiscze replied:
I don't remember them actually being displayed anywhere (correct me if i am wrong here), which is why i didn't add the field when reimplementing the records.

They can very easily be re-added though, if it is necessary for anything.

comma replied:
They're displayed in employment consoles, at the bottom. Yeah that's why I asked, seemed like an ez thing.

atlantiscze replied:
@comma If you don't mind, i'll add that in additional PR once this is merged, given the amount of affected files i'm pretty much spending all the time fixing conflicts at this point, so i'd rather have this merged as soon as possible.

comma commented on code/modules/mob/living/silicon/pai/software_modules.dm:
```diff
@@ -222,118 +220,6 @@
 		user.pda = new(user)
 		user.pda.set_owner_rank_job(text("[]", user), "Personal Assistant")
 
-/datum/pai_software/med_records
```
No plans to replace record viewer for pAIs? They can't really use modular computers.

atlantiscze replied:
They may be given access to the nano module in the future, just not doing it in this particular PR as it's touching way too much stuff at once already.

comma replied:
True

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2017-08-14-ShinyNewRecords

atlantiscze committed:
Map commit
